### ISSUE TITLE
UUID.properties() crashes because "0" is not a valid UUID

### ISSUE THREAD

**Issue by hashemi** [vapor/core#87]
This code crashes in the beta branch:

```swift
import CodableKit
import Foundation

_ = UUID.properties()
```

Producing the following error:

```
Thread 1: Fatal error: 'try!' expression unexpectedly raised an error: Swift.DecodingError.dataCorrupted(Swift.DecodingError.Context(codingPath: [], debugDescription: "Attempted to decode UUID from invalid UUID string.", underlyingError: nil))
```

This also happens whenever `Decodable.properties()` is called on a type that has a UUID field.

This happens because `CodingKeyCollector` works by pretending to decode a type. When it encounters a `String`, it tries to decode `"0"`:

```swift
    func decode(_ type: String.Type) throws -> String {
        result.add(type: type, atPath: codingPath)
        return "0"
    }
```

I was able to temporarily fix the crasher by changing that function in `CodingKeyCollectorSingleValueDecoder` to:

```swift
    func decode(_ type: String.Type) throws -> String {
        result.add(type: type, atPath: codingPath)
        return "00000000-0000-0000-0000-000000000000"
    }
```

A slightly more robust solution is to add a `decode(_ type: UUID.Type) -> UUID` that handles that special case.

Still, this only fixes this one case and doesn't address the class of potential errors that arise from assuming that any `Decodable` can be successfully decoded with an arbitrary combination of values (`false`, `"0"`, `0`).

**Comment by SandorDobi** [comment]
Hi,
The problem is twofold the UUID is failing to decode, but if we do like this it will raise another bug with duplicating fields. 
The same error is exist with Date type field. It will not crash but it cause field duplication in properties() method call.

My ugly hack is the following. It fixes the UUID problem and the Date problem, but it is not a proper fix. The proper way is to refactor the properties array to Dictionary to avoid duplicating the fields

Change the following two methods in `CodingKeyCollectorSingleValueDecoder` to

```
func decode(_ type: Double.Type) throws -> Double {
        if result.properties.count == 0 {
            result.add(type: type, atPath: codingPath)
        } else {
            let typeName = "\(result.properties.last!.type)"
            if  typeName != "Date" {
                result.add(type: type, atPath: codingPath)
            }
        }
        return 0
    }

    func decode(_ type: String.Type) throws -> String {
        if result.properties.count == 0 {
            result.add(type: type, atPath: codingPath)
            return "0"
        } else {
            let typeName = "\(result.properties.last!.type)"
            if  typeName == "UUID" {
                return UUID.keyStringFalse.uuidString
            } else {
                result.add(type: type, atPath: codingPath)
                return "0"
            }
        }
}
 
```

**Comment by SandorDobi** [comment]
@tanner0101 I think it can be closed now, as swift 4.1 refactoring is in place.......
